src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java

```diff
@@ -196,71 +199,95 @@ private static Input parseArguments(String[] args) {
 
-        performUpload(formatToFileNamesWrapper, input);
+        OkHttpClient client = OkHttpClientUtils.createClient(input.validateSsl);
+        try {
+            performUpload(client, reportPatternManager, input);
+        } finally {
+            // we must shut down OkHttp as otherwise it will leave threads running and
+            // prevent JVM shutdown
+            client.dispatcher().executorService().shutdownNow();
+            client.connectionPool().evictAll();
+        }
     }
 
-    private static void performUpload(FormatToFileNamesWrapper formatToFileNamesWrapper, Input input) throws IOException, AgentOptionParseException {
-        String sessionId = openSession(input);
-        for (String format : formatToFileNamesWrapper.getAllAvailableFormats()) {
-            sendRequestForFormat(input, format, formatToFileNamesWrapper.getFilesForFormat(format), sessionId);
+    private static void performUpload(OkHttpClient client, ReportPatternManager reportPatternManager, Input input) throws IOException, AgentOptionParseException {
+        String sessionId = openSession(client, input);
+        for (String format : reportPatternManager.getAllUsedFormats()) {
+            Set<String> patternsForFormat = reportPatternManager.getPatternsForFormat(format);
+            if (patternsForFormat.isEmpty()) {
+                fail("The input file contains no patterns for [" + format + "]." +
+                        " Did you forget to specify files to upload for that format?");
+            }
+            sendRequestForFormat(client, input, format, patternsForFormat, sessionId);
         }
-        closeSession(input, sessionId);
+        closeSession(client, input, sessionId);
     }
 
-    private static String openSession(Input input) throws IOException {
+    private static String openSession(OkHttpClient client, Input input) throws IOException {
         HttpUrl.Builder builder = input.url.newBuilder()
                 .addPathSegments("api/projects").addPathSegment(input.project)
                 .addPathSegments("external-analysis/session")
                 .addQueryParameter("partition", input.partition);
 
-        HttpUrl url = builder.build();
+        // We track revision or branch:timestamp for the session as it should be the same for all uploads
+        if (input.commit != null) {
+            builder.addQueryParameter("revision", input.commit);
+        } else if (input.timestamp != null) {
+            builder.addQueryParameter("t", input.timestamp);
+        } else if (input.autoDetectedCommit) {
+            String commit = detectCommit();
+            if (commit == null) {
+                fail("Failed to automatically detect the commit. Please specify it manually via --commit or --timestamp");
+            }
+            builder.addQueryParameter("revision", commit);
+        } else {
+            builder.addQueryParameter("t", "master:HEAD");
```

Comment:
this assumes Git and a default branch name of `master`. Can we somehow make this work with any default branch name, e.g. `trunk` from SVN?

Reply:
I am not really sure, how we would do that. Do you have any suggestions?

If this requires a non-trivial change, I would prefer to move this to a separate issue, so that we can get the session support merged.

Reply:
I had a quick look at the TS code and I think just passing `"HEAD"` should choose the default branch. Can you quickly check if that works? If not, please create a separate issue :-) :+1:

Reply:
I adjusted the value to "HEAD" and the upload for our git project still works. 👍 
We might at some point need to test this for SVN and others. But, for now, this should do.